amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -345,6 +358,30 @@ void initialize() {
         }
     }
 
+    @VisibleForTesting
+    boolean waitAndCheckIfLeaseTableIsReady() throws InterruptedException,
+            DependencyException, ProvisionedThroughputException, InvalidStateException {
+        long waitTime = ThreadLocalRandom.current().nextLong(MIN_WAIT_TIME_FOR_LEASE_TABLE_CHECK_MILLIS, MAX_WAIT_TIME_FOR_LEASE_TABLE_CHECK_MILLIS);
+        long waitUntil = System.currentTimeMillis() + waitTime;
+
+        while (System.currentTimeMillis() < waitUntil && leaseRefresher.isLeaseTableEmpty()) {
+            // check every 3 seconds if lease table is still empty,
+            // to minimize contention between all workers bootstrapping at the same time
+            log.info("Lease table is still empty. Checking again in {} ms", LEASE_TABLE_CHECK_FREQUENCY_MILLIS);
+            Thread.sleep(LEASE_TABLE_CHECK_FREQUENCY_MILLIS);
+        }
+        return !leaseRefresher.isLeaseTableEmpty();
+    }
+
+    private void waitUntilHashRangeCovered() throws InterruptedException {
+
+        while (!leaderElectedPeriodicShardSyncManager.hashRangeCovered()) {
+            // wait until entire hash range is covered
+            log.info("Hash range is not covered yet. Checking again in {} ms", HASH_RANGE_COVERAGE_CHECK_FREQUENCY_MILLIS);
+            Thread.sleep(HASH_RANGE_COVERAGE_CHECK_FREQUENCY_MILLIS);
+        }
+    }
+
```

Comment:
This should be per stream check. Does leaderElectedPeriodicShardSyncManager have context for all streams? If not should be passed though this method.

Reply:
Currently this call is not in use: We decided to exclude this call in the initialize() code in line 335. But if we do need it, at this point, leaderElectedPeriodicShardSyncManager does have the context of all streams as it has the reference of streamToShardSyncTaskManagerMap. This hashRangeCovered has not been implemented. I think that's part of Josh's implementation.

Reply:
The hashrange check from shards is in scope of Josh's work. hashrange check from leasetable should be the periodic syncer's feature. let's add this to our SIM for later tracking. Also add a TODO here with SIM